### PR TITLE
Extract markdown parser in its own plugin.

### LINKED ISSUE #5770: Add IMarkdownParser to rendermime interfaces
Following the model of `ILatexTypesetter`, a `IMarkdownParser` interface could be added to `rendermime` with default implementation based on markedjs.
This would make easier to replace the default jupyterlab markdown renderer.

This is related to #272.

Cheers

### PR DESCRIPTION
**Pull request by fcollonval**
## References

Closes #7291
Fixes #5770

The idea is to isolate `marked` js. And to make `@jupyterlab/application` not depend on codemirror. Before this PR it is brought by the markdown parser as the way to highlight code block. This makes small remix of JupyterLab heavy in term of JS assets.

## Code changes

- Create a new token `IMarkdownParser` in `@jupyterlab/rendermime`
- The rendermime registry uses optionally that token
- Create a new package `@jupyterlab/markedparser-extension` providing a concrete implementation of the markdown parser
- Use a default markdown renderer `<pre>${source}</pre>` if no provider exists.

## User-facing changes

N/A

## Backwards-incompatible changes

Extend renderer interface to receive a markdown parser

### CODE REVIEW

**Comment by jtpio** [comment]
> And to make @jupyterlab/application not depend on codemirror.

Really nice, thanks!

Some CI failures will be expected because of the new package.

**Comment by jtpio** [comment]
Thanks @fcollonval, the diff looks good :+1: 

We could also add that change to the extension migration guide here: https://jupyterlab.readthedocs.io/en/latest/extension/extension_migration.html#api-breaking-changes

Most of the CI failures are expected since the new package has not been published yet:

```
error An unexpected error occurred: "https://registry.yarnpkg.com/@jupyterlab%2fmarkedparser-extension: Not found".
```

Which means that a follow-up to merging this PR would be to make a new `4.0.0ax` prerelease right after to fix the CI.

**Comment by fcollonval** [comment]
I added the documentation in the migration guide and in the token list.

I also made the new parser optional in `rendermime-interfaces` to ease the migration.

**Comment by fcollonval** [comment]
@jtpio would you mind having a second look at this one. If it is ok I'll then take care of the merge followed by a release

**Comment by jtpio** [comment]
Thanks @fcollonval.

The diff looks great and should be ready to go. But I wanted to first get the CI back to green (https://github.com/jupyterlab/jupyterlab/issues/12023 and https://github.com/jupyterlab/jupyterlab/issues/12049) to reduce the noise when making the release, in case we need to troubleshoot some follow-up issues.

**Comment by jtpio** [comment]
Looks like there is a typing issue left:

image

**Comment by jtpio** [comment]
Also I would have liked to get https://github.com/jupyterlab/jupyterlab/pull/11894 in the next `4.0.0ax` pre-release.

So after the CI is fixed, maybe we can merge https://github.com/jupyterlab/jupyterlab/pull/11894, then this PR, and then make a new release?

**Comment by fcollonval** [comment]
> So after the CI is fixed, maybe we can merge #11894, then this PR, and then make a new release?

@trungleduc needs also a release for #11969. But no problem to wait for #11894

**Review by jtpio — APPROVED** [review]
Looks good, thanks!

**Comment by jtpio** [comment]
> And to make @jupyterlab/application not depend on codemirror. Before this PR it is brought by the markdown parser as the way to highlight code block. This makes small remix of JupyterLab heavy in term of JS assets.

@fcollonval were you able to measure the size improvements (for reference and have a trace here)?

**Comment by fcollonval** [comment]
Currently on [quetz-frontend](https://github.com/mamba-org/quetz-frontend) the webpack bundle analyzer is (in production mode):

image

The total parsed size is 23.44MB in which 2.87MB comes from codemirror + yjs and 2.16MB comes from codemirror modes. Meaning about 20% of the bundle.

If this is merged, as Blueprint JS is dropped too. This will reduce the bundle by an additional 3.25MB + 1.61MB. Bringing the total application size down by 42%.

**Comment by jtpio** [comment]
Nice!

**Comment by jtpio** [comment]
@fcollonval looks like the `js-rendermime` test failure might be relevant?

https://github.com/jupyterlab/jupyterlab/runs/5235525209?check_suite_focus=true

image

**Comment by jtpio** [comment]
> looks like the `js-rendermime` test failure might be relevant?

Probably we need to update the tests to take into account the new markdown parser option.

File: docs/source/extension/extension_migration.rst [hunk 1]
```diff
@@ -44,6 +44,10 @@ bumped their major version (following semver convention):
    * Command ``Collapsible_Headings:Toggle_Collapse`` renamed to ``notebook:toggle-heading-collapse``.
    * Command ``Collapsible_Headings:Collapse_All`` renamed to ``notebook:collapse-all-headings``.
    * Command ``Collapsible_Headings:Expand_All`` renamed to ``notebook:expand-all-headings``.
+- ``@jupyterlab/rendermime`` from 3.x to 4.x
+  The markdown parser has been extracted to its own plugin ``@jupyterlab/markedparser-extension:plugin``
+  that provides a new token ``IMarkdownParser`` (defined in ``@jupyterlab/rendermime``).
+  Consequently the ``IRendererFactory.createRenderer`` has a new option ``markdownparser``.
```
Comment (jtpio) [inline docs/source/extension/extension_migration.rst:50]:
```suggestion
  Consequently the ``IRendererFactory.createRenderer`` has a new option ``markdownParser``.
```

**Comment by jtpio** [comment]
Looks like there are a couple of conflicts to fix.

Then we should be able to merge and release a new `4.0.0ax`.

**Comment by jtpio** [comment]
Thanks @fcollonval for the fixes.

Mind doing one more rebase, so we pick up some of the recent changes that landed on `master`?

It looks like this should finally be good to go now.

**Comment by fcollonval** [comment]
Merging and carrying out an new alpha release

**Comment by jtpio** [comment]
Thanks!

**Pull request merged**